### DISCUSSION TITLE
PixelShuffle in flax?

### DISCUSSION THREAD

**Discussion by dslisleedh** [google/flax#2899] [Q&A]
PyTorch offers the `Pixelshuffle` layer, which increases the spatial size of a feature map by moving the channel dimension to the spatial dimension. While Keras does not have an equivalent `Pixelshuffle` layer, it can perform a similar operation to PyTorch using the `tf.nn.depth_to_space` function and the `Lambda` layer.

However, I was unable to find any depth-to-space-like functions or Pixelshuffle-like layers in either JAX or Flax. Is there a particular reason for this difference, or have I overlooked something?

**Comment by cgarciae** [comment]
Hey @dslisleedh, we don't have this layer but I think you can easily implement it using the [einops](https://github.com/arogozhnikov/einops) library, its even in one of their [examples](https://github.com/arogozhnikov/einops#result-is-strictly-determined).

Here is an example of space-to-depth and depth-to-space for some 2D data:

```python
from einops import rearrange
import jax.numpy as jnp

x = jnp.arange(5 * 8 * 8 * 3).reshape(5, 8, 8, 3)

# Space to depth
x = rearrange(x, '... (h b1) (w b2) c -> ... h w (c b1 b2)', b1=2, b2=2)
print(x.shape) # (5, 4, 4, 12)

# Depth to space
x = rearrange(x, '... h w (c b1 b2) -> ... (h b1) (w b2) c', b1=2, b2=2)
print(x.shape) # (5, 8, 8, 3)
```

**Reply by dslisleedh** [reply]
Thanks for sharing codes @cgarciae.

However, if `Pixelshuffle` can be easily implemented in Einops, I think it can also be implemented in pure jax, so why not implement it in Flax like other frameworks? 

It's a bit annoying that every time I implement a SuperResolution model, I have to use Einops or copy the `Pixelshuffle` implementation from a previous project, so I was thinking "is there a reason for this, or am I missing something?" and guessing from your answer, I'm not missing something, so I was wondering if there is a reason for not implementing Pixelshuffle in flax, or if you might consider implementing it for convenience. 

I'm attaching the `pixel_shuffle_nd` implementation for n dimensions, along with the results. Thank you very much for your time.

```python
import jax
import jax.numpy as jnp
import numpy as np

from absl.testing import parameterized, absltest

from typing import Sequence

jax.config.parse_flags_with_absl()


def pixel_shuffle_nd(x: jnp.ndarray, scale: Sequence[int]):
    is_single_input = (len(x.shape) - 2) != len(scale)
    if is_single_input:
        x = jnp.expand_dims(x, axis=0)

    spatial_size = x.shape[1:-1]
    assert len(spatial_size) == len(scale)
    c = x.shape[-1]
    n_div = np.prod(scale)
    assert c % n_div == 0

    x = jnp.reshape(x, (-1,) + spatial_size + scale + (c // n_div,))
    sort_idx = sum([(i, i + len(scale)) for i in range(1, len(scale) + 1)], ())
    x = jnp.transpose(x, (0,) + sort_idx + (len(sort_idx) + 1,))
    x = jnp.reshape(x, (-1,) + tuple(i * j for i, j in zip(spatial_size, scale)) + (c // n_div,))

    if is_single_input:
        x = jnp.squeeze(x, axis=0)
    return x


class PSTest(parameterized.TestCase):

    @parameterized.parameters(
        {'x': (1, 64, 64, 32), 'scale': (2, 2), 'y': (1, 128, 128, 8)},
        {'x': (64, 64, 32), 'scale': (2, 2), 'y': (128, 128, 8)},
        {'x': (16, 64, 64, 64, 32), 'scale': (2, 2, 2), 'y': (16, 128, 128, 128, 4)},
        {'x': (64, 64, 64, 128), 'scale': (4, 4, 4), 'y': (256, 256, 256, 2)},
    )
    def test_pixel_shuffle_nd(self, x, scale, y):
        x = jnp.ones(x)
        y = jnp.ones(y)
        y_hat = pixel_shuffle_nd(x, scale)
        self.assertEqual(y_hat.shape, y.shape)

        y_hat = jax.jit(pixel_shuffle_nd, static_argnums=(1,))(x, scale)
        self.assertEqual(y_hat.shape, y.shape)


if __name__ == '__main__':
    absltest.main()
```


```
(flax) dslisleedh@new-gpu:~/SASR-flax$ CUDA_VISIBLE_DEVICES=7 python test.py
Running tests under Python 3.10.0: /home/dslisleedh/miniconda3/envs/flax/bin/python
[ RUN      ] PSTest.test_pixel_shuffle_nd0 (x=(1, 64, 64, 32), scale=(2, 2), y=(1, 128, 128, 8))
I0224 10:54:23.957942 140706737104704 xla_bridge.py:353] Unable to initialize backend 'tpu_driver': NOT_FOUND: Unable to find driver in registry given worker: 
I0224 10:54:24.229936 140706737104704 xla_bridge.py:353] Unable to initialize backend 'rocm': NOT_FOUND: Could not find registered platform with name: "rocm". Available platform names are: CUDA Interpreter Host
I0224 10:54:24.230441 140706737104704 xla_bridge.py:353] Unable to initialize backend 'tpu': module 'jaxlib.xla_extension' has no attribute 'get_tpu_client'
I0224 10:54:24.230624 140706737104704 xla_bridge.py:353] Unable to initialize backend 'plugin': xla_extension has no attributes named get_plugin_device_client. Compile TensorFlow with //tensorflow/compiler/xla/python:enable_plugin_device set to true (defaults to false) to enable this.
[       OK ] PSTest.test_pixel_shuffle_nd0 (x=(1, 64, 64, 32), scale=(2, 2), y=(1, 128, 128, 8))
[ RUN      ] PSTest.test_pixel_shuffle_nd1 (x=(64, 64, 32), scale=(2, 2), y=(128, 128, 8))
[       OK ] PSTest.test_pixel_shuffle_nd1 (x=(64, 64, 32), scale=(2, 2), y=(128, 128, 8))
[ RUN      ] PSTest.test_pixel_shuffle_nd2 (x=(16, 64, 64, 64, 32), scale=(2, 2, 2), y=(16, 128, 128, 128, 4))
[       OK ] PSTest.test_pixel_shuffle_nd2 (x=(16, 64, 64, 64, 32), scale=(2, 2, 2), y=(16, 128, 128, 128, 4))
[ RUN      ] PSTest.test_pixel_shuffle_nd3 (x=(64, 64, 64, 128), scale=(4, 4, 4), y=(256, 256, 256, 2))
[       OK ] PSTest.test_pixel_shuffle_nd3 (x=(64, 64, 64, 128), scale=(4, 4, 4), y=(256, 256, 256, 2))
----------------------------------------------------------------------
Ran 4 tests in 1.727s

OK
```

**Reply by cgarciae** [reply]
@dslisleedh I understand, on Flax's side we usually only maintain the most basic building blocks and even then they are sometimes advertised as a reference implementations people can fork if needed. 

We don't have a "Addons" repo like Tensorflow does to include more specialized layers. For now I would recommend you create a separate repo with reusable Modules if you need to share them across projects.

That said, I will discuss this issue with the team as we've not thought about this problem.

**Reply by dslisleedh** [reply]
Maybe I've misunderstood Flax's philosophy. Thank you for considering my suggestion, @cgarciae.

**Reply by borisdayma** [reply]
I kind of agree with @dslisleedh that PixelShuffle/Unshuffle have become basic building blocks for vision.
They seem to be in most frameworks so would be convenient to have them natively eventually…

**Reply by cgarciae** [reply]
Interesting, thanks @borisdayma for the input!
I'll discuss it will the team :)

**Reply by jheek** [reply]
I think this pixel shuffle op isn't really a NN layer as it doesn't require any learned parameters or state. If this is difficult to do or very common I would say this belongs in jax.nn

**Reply by borisdayma** [reply]
Yeah, it's pretty straightforward to implement.
The main reason to have it in flax or jax would be if there's a specific trick making compilation or sharding more efficient.
If you shard across the embed dimension,  it's not necessarily obvious how your should rearrange the channels.
